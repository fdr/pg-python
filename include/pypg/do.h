#ifndef PyPg_do_H
#define PyPg_do_H 0
#ifdef __cplusplus
extern "C" {
#endif

Datum pl_inline(PG_FUNCTION_ARGS);

#ifdef __cplusplus
}
#endif
#endif /* !PyPg_do_H */
