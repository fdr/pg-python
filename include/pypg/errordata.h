/*
 * Postgres ErrorData interfaces
 */
#ifndef PyPg_errordata_H
#define PyPg_errordata_H 0
#ifdef __cplusplus
extern "C" {
#endif

typedef struct PyPgErrorData {
	PyObject_HEAD
	ErrorData *errordata;
} * PyPgErrorData;
extern PyTypeObject PyPgErrorData_Type;

PyObj PyPgErrorData_Initialize(PyObj self, ErrorData *ed);
PyObj PyPgErrorData_FromCurrent(void); /* CopyErrorData(); FlushErrorData(); */

#define PyPgErrorData_NEW()  \
	((PyObj) PyPgErrorData_Type.tp_alloc(&PyPgErrorData_Type, 0))
#define PyPgErrorData_New(ED) PyPgErrorData_Initialize(PyPgErrorData_NEW(&PyPgErrorData_Type), ED)

#define PyPgErrorData_Check(SELF) (PyObject_TypeCheck((SELF), &PyPgErrorData_Type))
#define PyPgErrorData_CheckExact(SELF) (Py_TYPE(SELF) == &PyPgErrorData_Type)

#define PyPgErrorData_GetErrorData(SELF) \
	(((PyPgErrorData)(SELF))->errordata)

#define PyPgErrorData_SetErrorData(SELF, ED) \
	(((PyPgErrorData)(SELF))->errordata) = ED

#ifdef __cplusplus
}
#endif
#endif /* !PyPg_errordata_H */
