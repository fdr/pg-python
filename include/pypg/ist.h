/*
 * pl IST management - routines used by Postgres.Transaction objects.
 */
#ifndef PyPg_ist_H
#define PyPg_ist_H 0
#ifdef __cplusplus
extern "C" {
#endif

extern unsigned long pl_ist_count; /* The current "xid". */

/* IST states */
#define pl_ist_open 'O'
#define pl_ist_committed 'C'
#define pl_ist_aborted 'X'
#define pl_ist_new ' '

/*
 * Aborts the current transaction.
 * Provides appropriate errors when given inappropriate state or xid.
 *
 * ON ERROR: Returns false and sets a Python exception.
 */
bool pl_ist_abort(unsigned long xid, char state);


/*
 * Commits the current transaction iff !pl_db_in_error; otherwise
 * abort the transaction and error.
 * Provides appropriate errors when given inappropriate state or xid.
 *
 * ON ERROR: Returns false and sets a Python exception.
 */
bool pl_ist_commit(unsigned long xid, char state);

/*
 * Start a new IST. Increments pl_ist_count by one.
 * Provides appropriate errors when given inappropriate state.
 *
 * ON ERROR: Returns false and sets a Python exception.
 */
bool pl_ist_begin(char state);

/*
 * Used in cases were the stored_ist_count != pl_ist_count
 *
 * ON ERROR: Postgres elog
 */
void pl_ist_reset(unsigned long n_to_abort);

#ifdef __cplusplus
}
#endif
#endif /* !PyPg_ist_H */
