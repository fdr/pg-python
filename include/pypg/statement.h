/*
 * Postgres prepared statements
 */
#ifndef PyPg_statement_H
#define PyPg_statement_H 0
#ifdef __cplusplus
extern "C" {
#endif

int execute_statements(const char *src);

typedef struct PyPgStatement {
	PyObject_HEAD
	PyObj ps_string;				/* SQL statement string */
	PyObj ps_input;					/* TupleDesc */
	PyObj ps_output;				/* TupleDesc */
	PyObj ps_command;				/* PyUnicode, CommandTag of the statement */
	PyObj ps_parameters;			/* constant parameters; None if none */

	MemoryContext ps_memory;		/* Context for statement data */
	Oid *ps_parameter_types;		/* Type Oids of the statement's parameters */
	OverrideSearchPath *ps_path;	/* Search path when the statement was created */
	SPIPlanPtr ps_plan;				/* Saved NO SCROLL Plan */
	SPIPlanPtr ps_scroll_plan;		/* Saved SCROLL Plan */
} * PyPgStatement;

extern PyTypeObject PyPgStatement_Type;
PyObj PyPgStatement_NEW(PyTypeObject *subtype,
	PyObj sql_statement_string, PyObj parameters);
SPIPlanPtr PyPgStatement_GetPlan(PyObj self);
SPIPlanPtr PyPgStatement_GetScrollPlan(PyObj self);

#define PyPgPreparedStatment_New(STATEMENT_STRING) PyPgStatement_NEW(&PyPgStatement_Type, STATEMENT_STRING)

#define PyPgStatement(SELF) ((PyPgStatement) SELF)

#define PyPgStatement_Check(SELF) (PyObject_TypeCheck(SELF, &PyPgStatement_Type))
#define PyPgStatement_CheckExact(SELF) (Py_TYPE(SELF) == &PyPgStatement_Type)

#define PyPgStatement_GetString(SELF) (PyPgStatement(SELF)->ps_string)
#define PyPgStatement_GetInput(SELF) (PyPgStatement(SELF)->ps_input)
#define PyPgStatement_GetOutput(SELF) (PyPgStatement(SELF)->ps_output)
#define PyPgStatement_GetParameters(SELF) (PyPgStatement(SELF)->ps_parameters)
#define PyPgStatement_GetParameterTypes(SELF) (PyPgStatement(SELF)->ps_parameter_types)
#define PyPgStatement_GetCommand(SELF) (PyPgStatement(SELF)->ps_command)
#define PyPgStatement_GetMemory(SELF) (PyPgStatement(SELF)->ps_memory)
#define PyPgStatement_GetPath(SELF) (PyPgStatement(SELF)->ps_path)

#define PyPgStatement_SetString(SELF, STR) (PyPgStatement(SELF)->ps_string = STR)
#define PyPgStatement_SetInput(SELF, TD) (PyPgStatement(SELF)->ps_input = TD)
#define PyPgStatement_SetOutput(SELF, TD) (PyPgStatement(SELF)->ps_output = TD)
#define PyPgStatement_SetParameters(SELF, PARAMS) (PyPgStatement(SELF)->ps_parameters = PARAMS)
#define PyPgStatement_SetParameterTypes(SELF, PTYPES) (PyPgStatement(SELF)->ps_parameter_types = PTYPES)
#define PyPgStatement_SetCommand(SELF, CMD) (PyPgStatement(SELF)->ps_command = CMD)
#define PyPgStatement_SetMemory(SELF, MEM) (PyPgStatement(SELF)->ps_memory = MEM)
#define PyPgStatement_SetPath(SELF, PATH) (PyPgStatement(SELF)->ps_path = PATH)

#define PyPgStatement_ReturnsRows(SELF) \
	(PyPgStatement_GetOutput(SELF) != Py_None)

#ifdef __cplusplus
}
#endif
#endif /* !PyPg_statement_H */
