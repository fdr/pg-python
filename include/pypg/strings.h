/*
 * Collection X-Macro's for building PyUnicode objects.
 */

#define PYPG_REQUISITE_NAMES() \
	IDSTR(exec) \
	IDSTR(__func__) \
	IDSTR(Postgres) \
	IDSTR(pg_errordata) \
	IDSTR(pg_inhibit_pl_context)

#define PYPG_ENTRY_POINTS() \
	IDSTR(main) \
	IDSTR(send) \
	IDSTR(before_insert) \
	IDSTR(before_update) \
	IDSTR(before_delete) \
	IDSTR(after_insert) \
	IDSTR(after_update) \
	IDSTR(after_delete) \
	IDSTR(before_insert_statement) \
	IDSTR(before_update_statement) \
	IDSTR(before_delete_statement) \
	IDSTR(before_truncate_statement) \
	IDSTR(after_insert_statement) \
	IDSTR(after_update_statement) \
	IDSTR(after_delete_statement) \
	IDSTR(after_truncate_statement)

#define TRIGGER_ORIENTATIONS() \
	IDSTR(ROW) \
	IDSTR(STATEMENT)

#define TRIGGER_TIMINGS() \
	IDSTR(BEFORE) \
	IDSTR(AFTER)

/*
 * Initialized into globals by _PG_init:
 *  INSERT_str_ob, UPDATE_str_ob, DELETE_str_ob, and TRUNCATE_str_ob.
 */
#define MANIPULATIONS() \
	IDSTR(INSERT) \
	IDSTR(UPDATE) \
	IDSTR(DELETE) \
	IDSTR(TRUNCATE)
