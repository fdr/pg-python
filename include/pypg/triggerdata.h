/*
 * Postgres TriggerData interface
 *
 * This isn't actually a full interface to TriggerData;
 * rather it's a collection of Python objects that represent the source
 * TriggerData.
 */
#ifndef PyPg_triggerdata_H
#define PyPg_triggerdata_H 0
#ifdef __cplusplus
extern "C" {
#endif

struct PyPgTriggerData {
	PyObject_HEAD
	PyObj td_reltype;			/* contains all of the table information */
	PyObj td_name;				/* trigger name */
	PyObj td_args;				/* tuple of PyUnicode */
	PyObj td_manipulation;	/* INSERT | UPDATE | DELETE | TRUNCATE */
	PyObj td_timing;			/* BEFORE | AFTER */
	PyObj td_orientation;	/* STATEMENT | ROW */
};
typedef struct PyPgTriggerData *PyPgTriggerData;
extern PyTypeObject PyPgTriggerData_Type;

PyObj PyPgTriggerData_Initialize(PyObj self, TriggerData *ed);
PyObj PyPgTriggerData_New(TriggerData *td);

#define PyPgTriggerData_NEW(TYP) \
	((PyObj) TYP.tp_alloc(&TYP, 0))
#define PyPgTriggerData_New(TD) PyPgTriggerData_Initialize(PyPgTriggerData_NEW(PyPgTriggerData_Type), TD)

#define PyPgTriggerData_Check(SELF) (PyObject_TypeCheck((SELF), &PyPgTriggerData_Type))
#define PyPgTriggerData_CheckExact(SELF) (Py_TYPE(SELF) == &PyPgTriggerData_Type)

#define PyPgTriggerData_GetRelationType(TD) \
	(((PyPgTriggerData) TD)->td_reltype)
#define PyPgTriggerData_GetTiming(TD) \
	(((PyPgTriggerData) TD)->td_timing)
#define PyPgTriggerData_GetManipulation(TD) \
	(((PyPgTriggerData) TD)->td_manipulation)

#ifdef __cplusplus
}
#endif
#endif /* !PyPg_triggerdata_H */
