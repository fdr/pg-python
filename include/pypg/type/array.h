/*
 * Postgres ArrayType
 */
#ifndef PyPg_type_array_H
#define PyPg_type_array_H 0
#ifdef __cplusplus
extern "C" {
#endif

#define PyPg_ARRAY_Type_oid InvalidOid
extern PyPgTypeObject PyPgArray_Type;
PyObj PyPgArray_Initialize(PyObj, ArrayType *);

#define PyPgArray_NEW() \
	(PyObj) PyPgArray_Type.tp_alloc(&PyPgArray_Type, 0)
#define PyPgArray_New(R) \
	PyPgArray_Initialize(PyPgArray_NEW(), R)

#define PyPgArray_Check(SELF) (PyObject_TypeCheck(SELF, &PyPgArray_Type))
#define PyPgArray_CheckExact(SELF) (Py_TYPE(SELF) == &PyPgArray_Type)

#define PyPgArray_GetArrayType(SELF) \
	((ArrayType *) (DatumGetPointer(PyPgObject_GetDatum(SELF))))

#define PyPgArray_SetArrayType(SELF, R) \
	PyPgObject_SetDatum(SELF, PointerGetDatum(R))

#ifdef __cplusplus
}
#endif
#endif /* !PyPg_type_array_H */
