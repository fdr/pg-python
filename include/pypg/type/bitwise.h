/*
 * Interface to "bitwise" types
 */
#ifndef PyPg_type_bitwise_H
#define PyPg_type_bitwise_H 0
#ifdef __cplusplus
extern "C" {
#endif

extern PyPgTypeObject PyPg_bool_Type;
#define PyPg_bool_Type_oid BOOLOID
extern PyPgTypeObject PyPg_bit_Type;
#define PyPg_bit_Type_oid BITOID
extern PyPgTypeObject PyPg_varbit_Type;
#define PyPg_varbit_Type_oid VARBITOID
extern PyPgTypeObject PyPg_bytea_Type;
#define PyPg_bytea_Type_oid BYTEAOID

#define PyPg_bool_Check(SELF) \
	(PyObject_TypeCheck((SELF), (PyTypeObject *) &PyPg_bool_Type))
#define PyPg_bool_CheckExact(SELF) (Py_TYPE(SELF) == &PyPg_bool_Type)

#define PyPg_bit_Check(SELF) \
	(PyObject_TypeCheck((SELF), (PyTypeObject *) &PyPg_bit_Type))
#define PyPg_bit_CheckExact(SELF) (Py_TYPE(SELF) == &PyPg_bit_Type)

#define PyPg_varbit_Check(SELF) \
	(PyObject_TypeCheck((SELF), (PyTypeObject *) &PyPg_varbit_Type))
#define PyPg_varbit_CheckExact(SELF) (Py_TYPE(SELF) == &PyPg_varbit_Type)

#define BYTES_RETURNS_OBJECT \
	XMETH(isspace) \
	XMETH(isalpha) \
	XMETH(isalnum) \
	XMETH(isdigit) \
	XMETH(islower) \
	XMETH(isupper) \
	XMETH(istitle)

#define BYTES_FILLS_OBJECT \
	XMETH(lower) \
	XMETH(upper) \
	XMETH(title) \
	XMETH(capitalize) \
	XMETH(swapcase)

#define PyPg_bytea_Check(SELF) \
	(PyObject_TypeCheck((SELF), (PyTypeObject *) &PyPg_bytea_Type))
#define PyPg_bytea_CheckExact(SELF) (Py_TYPE(SELF) == &PyPg_bytea_Type)

#ifdef __cplusplus
}
#endif
#endif /* !PyPg_type_bitwise_H */
