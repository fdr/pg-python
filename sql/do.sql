DO $$
import Postgres
sqlexec("""
	CREATE TABLE created_from_do (i int);
	INSERT INTO created_from_do VALUES (1004);
""")
$$ LANGUAGE python;
SELECT * FROM created_from_do;

DO $$
import Postgres
Postgres.ERROR("do-raised-error")
$$ LANGUAGE python;

DO $$
import Postgres
x = xact()
x.__enter__()
$$ LANGUAGE python;

-- Now for something more interesting...
DO $$
prepare("""
DO $more$
import Postgres
Postgres.WARNING("sub-do")
$more$ LANGUAGE python;
""")()
$$ LANGUAGE python;

-- Now for something even more interesting...
DO $$
prepare("""
DO $more$
import Postgres
Postgres.ERROR("how does the traceback look?")
$more$ LANGUAGE python;
""")()
$$ LANGUAGE python;
