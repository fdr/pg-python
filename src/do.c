/*
 * do.c - inline handler implementation
 *
 * Feature found in PG 8.5 and greater.
 */
#include <setjmp.h>

#define PY_SSIZE_T_CLEAN
#include <Python.h>
#include <compile.h>
#include <structmember.h>

#include "postgres.h"
#include "fmgr.h"
#include "funcapi.h"
#include "libpq/libpq-be.h"
#include "miscadmin.h"
#include "access/htup.h"
#include "access/heapam.h"
#include "access/xact.h"
#include "access/transam.h"
#include "catalog/namespace.h"
#include "catalog/pg_class.h"
#include "catalog/pg_database.h"
#include "catalog/pg_proc.h"
#include "catalog/pg_type.h"
#include "catalog/pg_language.h"
#include "catalog/indexing.h"
#include "storage/block.h"
#include "storage/off.h"
#include "commands/trigger.h"
#include "executor/spi.h"
#include "nodes/memnodes.h"
#include "tcop/tcopprot.h"
#include "utils/memutils.h"
#include "utils/array.h"
#include "utils/datum.h"
#include "utils/elog.h"
#include "utils/builtins.h"
#include "utils/hsearch.h"
#include "utils/syscache.h"
#include "utils/relcache.h"
#include "utils/typcache.h"
#include "mb/pg_wchar.h"

#include "pypg/environment.h"
#include "pypg/python.h"
#include "pypg/postgres.h"
#include "pypg/strings.h"
#include "pypg/externs.h"
#include "pypg/pl.h"
#include "pypg/do.h"
#include "pypg/errordata.h"
#include "pypg/triggerdata.h"
#include "pypg/errcodes.h"
#include "pypg/error.h"
#include "pypg/ist.h"
#include "pypg/type/type.h"
#include "pypg/type/object.h"
#include "pypg/type/record.h"
#include "pypg/type/array.h"
#include "pypg/type/bitwise.h"
#include "pypg/type/numeric.h"
#include "pypg/type/string.h"
#include "pypg/type/system.h"
#include "pypg/type/timewise.h"
#include "pypg/function.h"
#include "pypg/tupledesc.h"
#include "pypg/statement.h"
#include "pypg/cursor.h"
#include "pypg/module.h"
#include "pypg/xact.h"


/*
 * The sub-executor for inline functions: DO '...' LANGUAGE plpython3u;
 *
 * Fake fcinfo, flinfo, and fn_info. Inline execution goes through
 * the pl_handler.
 */
struct pl_fn_info pl_inline_executor_info;
FmgrInfo pl_inline_executor_fmgrinfo;
FunctionCallInfoData pl_inline_executor_fcinfo = {NULL,};

/*
 * Create the necessary function information for executing an inline procedure.
 * The executor is actually implemented in pure-Python in module.py, this
 * initialization path constructs the function information for that executor.
 *
 * By taking this route, we execute inline statements in the same context as
 * other functions, which allows us to leverage the PL-state checking already
 * available in plpython3_handler().
 */
static void
initialize_inline_executor(Oid langOid)
{
	MemoryContext former = CurrentMemoryContext;
	const char *attname = "prosrc";
	Oid cstringoid = CSTRINGOID;
	TupleDesc argdesc = NULL;
	PyObj func, module = NULL, output;
	PyObj fn_oid_int, fn_oid_str;
	PyObj nspname_str_ob, filename_str_ob;

	/*
	 * The fake function object will take a single cstring argument and return
	 * VOID.
	 */
	func = PyPgFunction_NEW();
	if (func == NULL)
	{
		PyErr_ThrowPostgresError("could not create the inline executor");
	}

	PyPgFunction_SetOid(func, InvalidOid);
	PyPgFunction_SetLanguage(func, langOid);
	PyPgFunction_SetNamespace(func, InvalidOid);
	PyPgFunction_SetStateful(func, false);
	PyPgFunction_SetXMin(func, InvalidTransactionId);
	PyPgFunction_SetReturnsSet(func, false);
	PyPgFunction_SetVolatile(func, 'v');
	PyPgFunction_SetPGFunction(func, NULL);

	Py_INCREF(Py_None);
	PyPgFunction_SetSource(func, Py_None);

	fn_oid_int = PyLong_FromLong(0);
	fn_oid_str = PyObject_Str(fn_oid_int);
	PyPgFunction_SetPyLongOid(func, fn_oid_int);
	PyPgFunction_SetPyUnicodeOid(func, fn_oid_str);

	filename_str_ob = PyUnicode_FromString("__inline_executor__");
	PyPgFunction_SetFilename(func, filename_str_ob);

	nspname_str_ob = PyUnicode_FromString("");
	PyPgFunction_SetNamespaceName(func, nspname_str_ob);

	/*
	 * Looking up types, the inline executor can't be initialized before first
	 * call.
	 */
	PG_TRY();
	{
		PyObj input;

		argdesc = TupleDesc_FromNamesAndOids(1, &attname, &cstringoid);
		input = PyPgTupleDesc_FromCopy(argdesc);
		PyPgFunction_SetInput(func, input);
		pl_inline_executor_info.fi_input = input;

		FreeTupleDesc(argdesc);
	}
	PG_CATCH();
	{
		PyErr_SetPgError(true);
	}
	PG_END_TRY();
	MemoryContextSwitchTo(former);

	output = PyPgType_FromOid(VOIDOID);
	PyPgFunction_SetOutput(func, output);

	module = PyObject_GetAttrString(Py_Postgres_module, "InlineExecutor");
	if (PyErr_Occurred())
	{
		Py_DECREF(func);
		Py_XDECREF(module);
		PyErr_ThrowPostgresError("could not initialize the inline executor");
	}

	pl_inline_executor_info.fi_xid = 1;
	pl_inline_executor_info.fi_func = func;
	/*
	 * The fake module is the primary contrast between the inline executor and
	 * real functions. The executor's fn_info is permanent and points to the
	 * Postgres.InlineExecutor object that provides a "main" attribute that will
	 * create and manage the temporary module to run the script in.
	 *
	 * This is how the inline executor avoids the usual initialize() procedure
	 * that real functions go through. It's *always* current.
	 */
	pl_inline_executor_info.fi_module = module;
	pl_inline_executor_info.fi_state = NULL;
	pl_inline_executor_info.fi_output = output;

	pl_inline_executor_fmgrinfo.fn_extra = &pl_inline_executor_info;
	pl_inline_executor_fmgrinfo.fn_oid = InvalidOid;
	pl_inline_executor_fmgrinfo.fn_addr = NULL;
	pl_inline_executor_fmgrinfo.fn_nargs = 1;
	pl_inline_executor_fmgrinfo.fn_strict = false;
	pl_inline_executor_fmgrinfo.fn_retset = false;
	pl_inline_executor_fmgrinfo.fn_stats = '\0';
	pl_inline_executor_fmgrinfo.fn_expr = NULL;

	pl_inline_executor_fcinfo.context = NULL;
	pl_inline_executor_fcinfo.resultinfo = NULL;
	pl_inline_executor_fcinfo.isnull = false;
	pl_inline_executor_fcinfo.nargs = 1;
	pl_inline_executor_fcinfo.flinfo = &pl_inline_executor_fmgrinfo;
}

PG_FUNCTION_INFO_V1(pl_inline);
Datum
pl_inline(PG_FUNCTION_ARGS)
{
	InlineCodeBlock *codeblock = (InlineCodeBlock *) DatumGetPointer(PG_GETARG_DATUM(0));
	Datum rd;

	Assert(IsA(codeblock, InlineCodeBlock));

	if (pl_inline_executor_fcinfo.flinfo == NULL)
	{
		if (pl_state == pl_not_initialized)
			pl_first_call();

		/*
		 * Initialize the pseudo-function for executing DO-blocks.
		 */
		initialize_inline_executor(codeblock->langOid);
	}

	/*
	 * Be sure to update the inline executor function if these
	 * parameters need to change.
	 */
	pl_inline_executor_fmgrinfo.fn_mcxt = CurrentMemoryContext;
	pl_inline_executor_fcinfo.argnull[0] = false;
	pl_inline_executor_fcinfo.arg[0] = PointerGetDatum(codeblock->source_text);

	rd = pl_handler(&pl_inline_executor_fcinfo);

	fcinfo->argnull[0] = true;
	fcinfo->arg[0] = PointerGetDatum(NULL);
	pl_inline_executor_fmgrinfo.fn_mcxt = NULL;

	return(rd);
}
