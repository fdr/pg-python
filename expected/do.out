DO $$
import Postgres
sqlexec("""
	CREATE TABLE created_from_do (i int);
	INSERT INTO created_from_do VALUES (1004);
""")
$$ LANGUAGE python;
SELECT * FROM created_from_do;
  i   
------
 1004
(1 row)

DO $$
import Postgres
Postgres.ERROR("do-raised-error")
$$ LANGUAGE python;
ERROR:  do-raised-error
CONTEXT:  [exception from Python]
Traceback (most recent call last):
   File "<string>", line 100, in main
   File "<string>", line 94, in load_module
   File "[do-block-2]", line 3, in <module>
    Postgres.ERROR("do-raised-error")
   File "<string>", line 49, in ERROR
 Postgres.Exception

[__inline_executor__]
DO $$
import Postgres
x = xact()
x.__enter__()
$$ LANGUAGE python;
ERROR:  function failed to exit all subtransactions
DETAIL:  The 1 remaining subtransactions have been aborted.
-- Now for something more interesting...
DO $$
prepare("""
DO $more$
import Postgres
Postgres.WARNING("sub-do")
$more$ LANGUAGE python;
""")()
$$ LANGUAGE python;
WARNING:  sub-do
CONTEXT:  SQL statement "
DO $more$
import Postgres
Postgres.WARNING("sub-do")
$more$ LANGUAGE python;
"
-- Now for something even more interesting...
DO $$
prepare("""
DO $more$
import Postgres
Postgres.ERROR("how does the traceback look?")
$more$ LANGUAGE python;
""")()
$$ LANGUAGE python;
ERROR:  how does the traceback look?
CONTEXT:  SQL statement "
DO $more$
import Postgres
Postgres.ERROR("how does the traceback look?")
$more$ LANGUAGE python;
"
[exception from Python]
Traceback (most recent call last):
   File "<string>", line 100, in main
   File "<string>", line 94, in load_module
   File "[do-block-7]", line 3, in <module>
    Postgres.ERROR("how does the traceback look?")
   File "<string>", line 49, in ERROR
 Postgres.Exception

[__inline_executor__]
[exception from Python]
Traceback (most recent call last):
   File "<string>", line 100, in main
   File "<string>", line 94, in load_module
   File "[do-block-6]", line 7, in <module>
    """)()
 Postgres.Exception

[__inline_executor__]
